### PR TITLE
Fix ExchangeClient not properly release ExchangeQueue issue upon close

### PR DESCRIPTION
**Pull request by tanjialiang**
When Task terminates and cleans up resource, close() is called for 
ExchangeClient. But close() omits to gracefully clean up resources in 
ExchangeQueue when there is a MemoryPool provided in SerializedPage.

Crash will happen when there is still unprocessed SerializedPage in 
ExchangeQueue in ExchangeClient when Task is destructed. This is because of the 
destruction order of fields in Task: MemoryPool is destructed before 
ExchangeClient, making ExchangeClient not able to properly use MemoryPool to 
cleanup upon the invocation of its destructor.

### CODE REVIEW

File: velox/exec/Exchange.cpp [hunk 1]
```diff
@@ -226,22 +210,27 @@ void ExchangeClient::noMoreRemoteTasks() {
...
   }
-
-  // Outside of mutex.
-  for (auto& source : sources) {
-    source->close();
+  {
+    std::shared_ptr<ExchangeQueue> queue;
+    queue = std::move(queue_);
+    VELOX_CHECK(queue.unique());
   }
```
Comment (tanjialiang) [inline velox/exec/Exchange.cpp:234]: This is the actual fix.

File: velox/exec/Exchange.cpp [hunk 2]
```diff
@@ -58,127 +60,110 @@ void SerializedPage::prepareStreamForDeserialize(ByteStream* input) {
...
-}
-
 // static
 std::vector<ExchangeSource::Factory>& ExchangeSource::factories() {
   static std::vector<Factory> factories;
   return factories;
 }
 
-namespace {
-class LocalExchangeSource : public ExchangeSource {
```
Comment (tanjialiang) [inline velox/exec/Exchange.cpp:82]: Exposing LocalExchangeSource to build a memory pool supported LocalExchangeSource for testing.

File: velox/exec/Exchange.h [hunk 3]
```diff
@@ -302,7 +331,7 @@ class ExchangeClient {
     return pool_;
   }
 
-  void maybeSetMemoryPool(memory::MemoryPool* pool);
+  void initialize(memory::MemoryPool* pool);
```
Comment (xiaoxmeng) [inline velox/exec/Exchange.h:334]: Add folly null annotation?

File: velox/exec/Exchange.cpp [hunk 4]
```diff
@@ -58,127 +60,110 @@ void SerializedPage::prepareStreamForDeserialize(ByteStream* input) {
...
+  return !pending;
+}
+
+void LocalExchangeSource::close() {
+  auto buffers = PartitionedOutputBufferManager::getInstance().lock();
+  buffers->deleteResults(taskId_, destination_);
+}
 
-void ExchangeClient::maybeSetMemoryPool(memory::MemoryPool* pool) {
+void ExchangeClient::initialize(memory::MemoryPool* pool) {
```
Comment (xiaoxmeng) [inline velox/exec/Exchange.cpp:166]: The comment is stale now. I am wondering if we need to make the change here in this PR?

File: velox/exec/Exchange.cpp [hunk 5]
```diff
@@ -226,22 +210,22 @@ void ExchangeClient::noMoreRemoteTasks() {
...
 void ExchangeClient::close() {
-  std::vector<std::shared_ptr<ExchangeSource>> sources;
-
   {
-    std::lock_guard<std::mutex> l(queue_->mutex());
-
-    if (closed_) {
-      return;
+    std::vector<std::shared_ptr<ExchangeSource>> sources;
+    {
```
Comment (xiaoxmeng) [inline velox/exec/Exchange.cpp:215]: This is not safe as the source could still add new page into queue_ after we release the lock and before we call source close. Consider
```
std::vector<std::shared_ptr<ExchangeSource>> sources;
{
   std::lock_guard<std::mutex> l(queue_->mutex());
   if (closed_) {
      return;
   }
   closed_ = true;
   sources = std::move(sources_);
}
// Close the sources outside of queue lock to avoid deadlock.
for (auto& source : sources) {
   source->close();
}
// Close the queue by discarding any buffered pages after we close all the sources.
{
   std::lock_guard<std::mutex> l(queue_->mutex());
   queue_->closeLock();
}
```

File: velox/exec/Exchange.h [hunk 6]
```diff
@@ -171,6 +171,11 @@ class ExchangeQueue {
     checkComplete();
   }
 
+  void closeLocked() {
+    std::deque<std::unique_ptr<SerializedPage>> queue;
+    queue = std::move(queue_);
```
Comment (xiaoxmeng) [inline velox/exec/Exchange.h:176]: Can we just call queue_.clear() any consideration for the move? Thanks!

**Review by xiaoxmeng — COMMENTED** [review]
Thanks for the change and have some comments.

File: velox/exec/Exchange.h [hunk 7]
```diff
@@ -209,20 +214,26 @@ class ExchangeSource : public std::enable_shared_from_this<ExchangeSource> {
   using Factory = std::function<std::shared_ptr<ExchangeSource>(
       const std::string& taskId,
       int destination,
-      std::shared_ptr<ExchangeQueue> queue)>;
+      std::shared_ptr<ExchangeQueue> queue,
+      memory::MemoryPool* pool)>;
```
Comment (mbasmanova) [inline velox/exec/Exchange.h:218]: @tanjialiang This is a nice change as we discussed earlier. Let's move it into a separate PR to make this PR smaller and easier to review.

File: velox/exec/Exchange.h (hunk 6 above)
Comment (mbasmanova) [reply velox/exec/Exchange.h:176]: +1

File: velox/exec/tests/MultiFragmentTest.cpp [hunk 8]
```diff
@@ -787,3 +788,147 @@ TEST_F(MultiFragmentTest, earlyCompletionMerge) {
...
+      return std::make_unique<CustomSlowOperator>(id, ctx, slowNode);
+    }
+    return nullptr;
+  }
+};
+
+TEST_F(MultiFragmentTest, exchangeDestruction) {
+  // This unit test tests the proper destruction of ExchangeClient upon
+  // task destruction.
+  OperatorTestBase::SetUp();
```
Comment (mbasmanova) [inline velox/exec/tests/MultiFragmentTest.cpp:876]: Should not be needed.

File: velox/exec/tests/MultiFragmentTest.cpp [hunk 9]
```diff
@@ -787,3 +788,147 @@ TEST_F(MultiFragmentTest, earlyCompletionMerge) {
...
+  Operator::registerOperator(std::make_unique<CustomSlowOperatorTranslator>());
+  ExchangeSource::factories().clear();
```
Comment (mbasmanova) [inline velox/exec/tests/MultiFragmentTest.cpp:878]: This will make it not possible to run tests in parallel. Consider using a different prefix for the custom exchange.

File: velox/exec/tests/MultiFragmentTest.cpp [hunk 10]
```diff
@@ -787,3 +788,147 @@ TEST_F(MultiFragmentTest, earlyCompletionMerge) {
...
+  ExchangeSource::registerFactory([](const std::string& taskId,
+                                     int destination,
+                                     std::shared_ptr<ExchangeQueue> queue,
+                                     memory::MemoryPool* pool) {
+    if (strncmp(taskId.c_str(), "local://", 8) == 0) {
+      return std::make_unique<LocalExchangeSource>(
```
Comment (mbasmanova) [inline velox/exec/tests/MultiFragmentTest.cpp:884]: This factory looks very similar to default one. What's the difference?

File: velox/exec/Task.h [hunk 11]
```diff
@@ -641,6 +641,16 @@ class Task : public std::enable_shared_from_this<Task> {
   /// node IDs specified in split management methods.
   const std::unordered_set<core::PlanNodeId> splitPlanNodeIds_;
 
+  std::unique_ptr<memory::MemoryPool> pool_;
```
Comment (mbasmanova) [inline velox/exec/Task.h:644]: It might be cleaner to move exchangClientXxx member variable. 

Consider adding a comment just above pool_ saying that all member variables that hold references to pool_ must be defined after pool_.

File: velox/exec/Exchange.h (hunk 7 above)
Comment (tanjialiang) [reply velox/exec/Exchange.h:218]: Hi @mbasmanova There is a PR already that has this change. After landing that PR and rebasing this one. The changes will be gone from this PR. I'm manually pasting the chagnes just to get this PR out.

File: velox/exec/Exchange.cpp (hunk 4 above)
Comment (tanjialiang) [reply velox/exec/Exchange.cpp:166]: The refactoring will also be appearing in another PR. After rebasing the changes here will disappear. I'm planning to land the other refactoring PR before this one.

File: velox/exec/tests/MultiFragmentTest.cpp (hunk 10 above)
Comment (tanjialiang) [reply velox/exec/tests/MultiFragmentTest.cpp:884]: This factory passes a pool but the default one pass nullptr to the pool param

File: velox/exec/tests/MultiFragmentTest.cpp (hunk 9 above)
Comment (tanjialiang) [reply velox/exec/tests/MultiFragmentTest.cpp:878]: Hi @mbasmanova could you explain more here? I am not sure if I understand it.

File: velox/exec/Exchange.h (hunk 7 above)
Comment (mbasmanova) [reply velox/exec/Exchange.h:217]: Got it. Found #2447.

File: velox/exec/tests/MultiFragmentTest.cpp (hunk 10 above)
Comment (mbasmanova) [reply velox/exec/tests/MultiFragmentTest.cpp:883]: Is there any reason for the default factor to pass null. Can we update the default to pass the pool instead (in the PR doing the refactoring)?

File: velox/exec/tests/MultiFragmentTest.cpp (hunk 9 above)
Comment (mbasmanova) [reply velox/exec/tests/MultiFragmentTest.cpp:877]: When multiple tests are running in parallel, tests that rely on default factory may fail either because the factory has been unregistered or because the new factory registered in its place has a different behavior.

File: velox/exec/Exchange.cpp (hunk 1 above)
Comment (spershin) [reply velox/exec/Exchange.cpp]: @tanjialiang 
This comment no longer actual as the code changed.

File: velox/exec/tests/MultiFragmentTest.cpp (hunk 10 above)
Comment (mbasmanova) [reply velox/exec/tests/MultiFragmentTest.cpp:889]: @tanjialiang Is it necessary to use a cluster Exchange Source? Looks like the default can work just fine, no?

File: velox/exec/tests/MultiFragmentTest.cpp [hunk 12]
```diff
@@ -787,3 +791,149 @@ TEST_F(MultiFragmentTest, earlyCompletionMerge) {
...
+
+    auto rootTask = makeTask("root-task", rootPlan, 0);
+    Task::start(rootTask, 1);
+    addRemoteSplits(rootTask, {leafTaskId});
+
+    waitForTaskCompletion(rootTask.get(), 1'000'000);
+    leafTask->requestAbort().get();
+    rootTask->requestAbort().get();
+  }
+  // Destructors of root Task should be called without issues. Unprocessed
```
Comment (mbasmanova) [inline velox/exec/tests/MultiFragmentTest.cpp:936]: Maybe destroy the task explicitly: `rootTask = nullptr;`

File: velox/exec/Exchange.cpp (hunk 2 above)
Comment (tanjialiang) [reply velox/exec/Exchange.cpp]: This is no longer needed.

File: velox/exec/Exchange.cpp [hunk 13]
```diff
@@ -156,6 +147,15 @@ class LocalExchangeSource : public ExchangeSource {
         });
   }
 
+  bool shouldRequestLocked() override {
```
Comment (mbasmanova) [inline velox/exec/Exchange.cpp:150]: Any particular reason to move this method?

File: velox/exec/Exchange.cpp [hunk 14]
```diff
@@ -222,22 +222,25 @@ void ExchangeClient::noMoreRemoteTasks() {
 }
 
 void ExchangeClient::close() {
-  std::vector<std::shared_ptr<ExchangeSource>> sources;
-
   {
```
Comment (mbasmanova) [inline velox/exec/Exchange.cpp:225]: Why do we need this extra scope?

File: velox/exec/Exchange.cpp (hunk 13 above)
Comment (tanjialiang) [reply velox/exec/Exchange.cpp:150]: The order is messed when I copy back the code. Let me change it

File: velox/exec/Task.h [hunk 15]
```diff
@@ -650,10 +650,6 @@ class Task : public std::enable_shared_from_this<Task> {
   // kFinished.
   bool partitionedOutputConsumed_ = false;
 
-  /// Exchange clients. One per pipeline / source.
-  /// Null for pipelines, which don't need it.
-  std::vector<std::shared_ptr<ExchangeClient>> exchangeClients_;
```
Comment (mbasmanova) [inline velox/exec/Task.h:655]: Should we move exchangeClientByPlanNode_ also to be next to exchangeClients_? It is a bit strange to have these far away.

File: velox/exec/tests/MultiFragmentTest.cpp [hunk 16]
```diff
@@ -787,3 +787,133 @@ TEST_F(MultiFragmentTest, earlyCompletionMerge) {
...
+      std::shared_ptr<const CustomSlowNode> slowNode)
+      : Operator(
+            driverCtx,
+            slowNode->outputType(),
+            operatorId,
+            slowNode->id(),
+            "CustomSlowOperator") {}
+
+  void addInput(RowVectorPtr input) override {
+    inputs_.emplace_back(std::move(input));
```
Comment (mbasmanova) [inline velox/exec/tests/MultiFragmentTest.cpp:828]: needsInput() returns false if there is at least one entry in 'inputs_', Hence, 'inputs_' will never have more than one entry. Hence, we don't need an std::vector for 'inputs_' and can store the input in 'input_' variable in the base class.

File: velox/exec/tests/MultiFragmentTest.cpp [hunk 17]
```diff
@@ -787,3 +787,133 @@ TEST_F(MultiFragmentTest, earlyCompletionMerge) {
...
+    return BlockingReason::kNotBlocked;
+  }
+
+  bool isFinished() override {
+    return false;
+  }
+
+ private:
+  std::vector<RowVectorPtr> inputs_;
+  bool consumedOneBatch{false};
```
Comment (mbasmanova) [inline velox/exec/tests/MultiFragmentTest.cpp:859]: naming: consumedOneBatch -> consumedOneBatch_

File: velox/exec/tests/MultiFragmentTest.cpp [hunk 18]
```diff
@@ -787,3 +787,133 @@ TEST_F(MultiFragmentTest, earlyCompletionMerge) {
     ASSERT_TRUE(waitForTaskCompletion(task.get())) << task->taskId();
   }
 }
+
+class CustomSlowNode : public core::PlanNode {
```
Comment (mbasmanova) [inline velox/exec/tests/MultiFragmentTest.cpp:791]: nit: We can shorten class names by removing "Custom".

File: velox/exec/tests/MultiFragmentTest.cpp [hunk 19]
```diff
@@ -787,3 +787,133 @@ TEST_F(MultiFragmentTest, earlyCompletionMerge) {
...
+                 .project({"c0 % 10 AS c0", "c1"})
+                 .partitionedOutput({}, 1)
+                 .planNode();
+
+  auto leafTask = makeTask(leafTaskId, leafPlan, 0);
+  Task::start(leafTask, 1);
+  addHiveSplits(leafTask, filePaths_);
+
+  core::PlanNodePtr rootPlan;
+  rootPlan = PlanBuilder()
```
Comment (mbasmanova) [inline velox/exec/tests/MultiFragmentTest.cpp:898]: nit: combine declaration with initialization

```
auto rootPlan = ...
```

File: velox/exec/tests/MultiFragmentTest.cpp [hunk 20]
```diff
@@ -787,3 +787,133 @@ TEST_F(MultiFragmentTest, earlyCompletionMerge) {
...
+                   return std::make_shared<CustomSlowNode>(id, std::move(node));
+                 })
+                 .partitionedOutput({}, 1)
+                 .planNode();
+
+  auto rootTask = makeTask("root-task", rootPlan, 0);
+  Task::start(rootTask, 1);
+  addRemoteSplits(rootTask, {leafTaskId});
+
+  waitForTaskCompletion(rootTask.get(), 1'000'000);
```
Comment (mbasmanova) [inline velox/exec/tests/MultiFragmentTest.cpp:910]: What do we expect to happen? I assume task will not complete. Should we assert it hasn't completed?

**Review by mbasmanova — COMMENTED** [review]
@tanjialiang Looks good to me % a couple small comments. Does the test fail or crash without the fix?

**Comment by tanjialiang** [comment]
> @tanjialiang Looks good to me % a couple small comments. Does the test fail or crash without the fix?

Yeah the test fail without the fix

File: velox/exec/tests/MultiFragmentTest.cpp [hunk 21]
```diff
@@ -787,3 +787,131 @@ TEST_F(MultiFragmentTest, earlyCompletionMerge) {
...
+  BlockingReason isBlocked(ContinueFuture* future) override {
+    return BlockingReason::kNotBlocked;
+  }
+
+  bool isFinished() override {
+    return false;
+  }
+
+ private:
+  RowVectorPtr input_;
```
Comment (mbasmanova) [inline velox/exec/tests/MultiFragmentTest.cpp:856]: This is not needed. This variable already exists in the base class.

**Review by mbasmanova — APPROVED** [review]
@tanjialiang Thank you for investigating and fixing this problem. Please, wrap PR description at 80 characters before landing.
